metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -0,0 +1,186 @@
+package io.iohk.metronome.checkpointing.service
+
+import cats.data.{NonEmptyList, NonEmptyVector, OptionT}
+import cats.effect.concurrent.{Ref, Semaphore}
+import cats.effect.{Concurrent, Resource, Sync}
+import cats.implicits._
+import io.iohk.metronome.checkpointing.CheckpointingAgreement
+import io.iohk.metronome.checkpointing.models.{
+  Block,
+  CheckpointCertificate,
+  Ledger
+}
+import io.iohk.metronome.checkpointing.service.storage.LedgerStorage
+import io.iohk.metronome.crypto.ECPublicKey
+import io.iohk.metronome.hotstuff.consensus.basic.{Phase, QuorumCertificate}
+import io.iohk.metronome.hotstuff.service.ApplicationService
+import io.iohk.metronome.hotstuff.service.storage.{
+  BlockStorage,
+  ViewStateStorage
+}
+import io.iohk.metronome.storage.KVStoreRunner
+
+class CheckpointingService[F[_]: Sync, N] private (
+    ledgerTree: Ref[F, Map[Block.Hash, (Ledger, Block.Header)]],
+    ledgerUpdateSemaphore: Semaphore[F],
+    maxConcurrentUpdates: Int,
+    ledgerStorage: LedgerStorage[N],
+    blockStorage: BlockStorage[N, CheckpointingAgreement]
+)(implicit storeRunner: KVStoreRunner[F, N])
+    extends ApplicationService[F, CheckpointingAgreement] {
+
+  override def createBlock(
+      highQC: QuorumCertificate[CheckpointingAgreement]
+  ): F[Option[Block]] = ???
+
+  override def validateBlock(block: Block): F[Boolean] = {
+    {
+      //TODO: Validate transactions PM-3131/3132
+      false.pure[F]
+    } >>
+      ledgerUpdateSemaphore.withPermit {
+        updateLedger(block).map {
+          _.hash == block.header.postStateHash
+        }
+      }
+  }
+
+  override def executeBlock(
+      block: Block,
+      commitQC: QuorumCertificate[CheckpointingAgreement],
+      commitPath: NonEmptyList[Block.Hash]
+  ): F[Unit] = {
+    require(commitQC.phase == Phase.Commit, "Commit QC required")
+    if (commitQC.blockHash != block.hash) {
+      //Several options here, depending on how we want to schedule the execution (BlockExecutor):
+      // - disallow this condition
+      // - ignore
+      // - execute block without constructing CC
+      ().pure[F]
+    } else {
+      for {
+        _          <- ledgerUpdateSemaphore.acquireN(maxConcurrentUpdates)
+        ledger     <- updateLedger(block)
+        checkpoint <- constructCheckpoint(ledger, commitQC)
+        _          <- saveCommittedLedger(ledger)
+        _          <- resetLedgerTree(ledger, block)
+        _          <- ledgerUpdateSemaphore.releaseN(maxConcurrentUpdates)
+        _          <- pushCheckpoint(checkpoint)
+      } yield ()
+    }
+  }
+
+  private def updateLedger(block: Block): F[Ledger] = {
+
+    // I'm worried that this may loop for a long time (until it reaches a pruned block)
+    // when the initial block is an ancestor of the last committed block - could that happen?
```

Comment:
Difficult to say.. I'm trying to get rid of a High Q.C. that has an earlier view than our own Commit Q.C. in the `ConsensusService`. I maintained the possibility in the `RobotService` and checked first if there's an ancestry relationship between the last executed block and the one I'm evaluating. 

I suppose it is possible because of concurrency: you may be asked to evaluate some other proposal block, but it blocks on the semaphore, and by the time it is allowed to go, another one fiber cleared the ledger tree, including the ancestors of this alternative proposal, which will now keep looping.

If you persisted the ledgers of all ancestor blocks, and also looked for it here as a fallback from the tree, wouldn't it feel you with more confidence that it will not loop back to genesis (and hold all blocks in memory along the way)?

Reply:
> If you persisted the ledgers of all ancestor blocks, and also looked for it here as a fallback from the tree

Sorry, I was wrong there, since the block to validate can skip some levels. 

I guess one thing that could help is to add a `height` field to the blocks, which could be used to eliminate traversing back more than our last executed block: if the block we are traversing has a lower height, we should have hit the root already, or we are on a different branch.

Not sure if this is important enough to be added to the `Block` type class itself and be supported in `BlockStorage`, but you can definitely do it in the `checkpointing.model.Block` and make use of it here.

Reply:
A side note: I think the lightest touch to add support for early stop to `BlockStorage.getPathFromAncestor` is to add a `maxDistance` parameter to the method. Since it's not retrieving blocks it cannot check the height, but if on the call site we know the height of our potential ancestor (the root of the state tree) and our new block, we know that if we had to go higher more than that, we should have found a path, or there isn't one.

Reply:
Added `maxDistance` in https://github.com/input-output-hk/metronome/pull/39/commits/907bbe32432865c5153004ff5abdb27b2da19d72 just in case.

Reply:
I've added block height and I'm using it not to traverse beyond last committed/executed block height.

Reply:
Good change! 

I thought my earlier attempt with `maxDistance` was a bit lacklustre; added `height` to the `Block` type class in https://github.com/input-output-hk/metronome/pull/55